lib/authorization/index.js

```diff
@@ -0,0 +1,292 @@
+    } else if (!secret && req.body) {
+      // try to get the secret from the body, but don't leave it there
+      if (_.isArray(req.body) && req.body.length > 0 && req.body[0].secret) {
+        secret = req.body[0].secret;
+        delete req.body[0].secret;
+      } else if (req.body.secret) {
+        secret = req.body.secret;
+        delete req.body.secret;
+      }
+    }
+
+    if (secret) {
+      // store the secret hash on the request since the req may get processed again
+      req.api_secret = secret;
+    }
+
+    return secret;
+  }
+
+ function authorizeAdminSecretWithRequest (req) {
+    return authorizeAdminSecret(adminSecretFromRequest(req));
+  }
+
+  function authorizeAdminSecret (secret) {
+    return (env.api_secret && env.api_secret.length > 12) ? (secret === env.api_secret) : false;
+  }
+
+  authorization.seenPermissions = [ ];
+
+  authorization.expandedPermissions = function expandedPermissions ( ) {
+    var permissions = shiroTrie.new();
+    permissions.add(authorization.seenPermissions);
+    return permissions;
+  };
+
+  authorization.resolveWithRequest = function resolveWithRequest (req, callback) {
+    authorization.resolve({
+      api_secret: adminSecretFromRequest(req)
+      , token: extractToken(req)
+    }, callback);
+  };
+
+  authorization.checkMultiple = function checkMultiple(permission, shiros) {
+    var found = _.find(shiros, function checkEach (shiro) {
+      return shiro && shiro.check(permission);
+    });
+    return _.isObject(found);
+  };
+
+  authorization.resolve = function resolve (data, callback) {
+
+    var defaultShiros = storage.rolesToShiros(defaultRoles);
+
+    if (storage.doesAccessTokenExist(data.api_secret)) {
+      authorization.resolveAccessToken (data.api_secret, callback, defaultShiros);
+      return;
+    }
+
+    if (authorizeAdminSecret(data.api_secret)) {
+      var admin = shiroTrie.new();
+      admin.add(['*']);
+      return callback(null, { shiros: [ admin ] });
+    }
+
+    if (data.token) {
+      jwt.verify(data.token, env.api_secret, function result(err, verified) {
+        if (err) {
+          return callback(err, { shiros: [ ] });
+        } else {
+          authorization.resolveAccessToken (verified.accessToken, callback, defaultShiros);
+        }
+      });
+    } else {
+      return callback(null, { shiros: defaultShiros });
+    }
+
+  };
+
+  authorization.resolveAccessToken = function resolveAccessToken (accessToken, callback, defaultShiros) {
+
+    if (!defaultShiros) {
+      defaultShiros = storage.rolesToShiros(defaultRoles);
+    }
+
+    let resolved = storage.resolveSubjectAndPermissions(accessToken);
+    if (!resolved || !resolved.subject) {
+      return callback('Subject not found', null);
+    }
+
+    let shiros = resolved.shiros.concat(defaultShiros);
+    return callback(null, { shiros: shiros, subject: resolved.subject });
+  };
+
+  authorization.isPermitted = function isPermitted (permission, opts) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['opts' is defined but never used (no-unused-vars)](https://app.codacy.com/manual/m-kozak/cgm-remote-monitor/pullRequest?prid=1196151)